Framework/Algorithms/src/CalMuonDetectorPhases.cpp

```diff
@@ -539,5 +545,27 @@ void CalMuonDetectorPhases::reportProgress(const int thisSpectrum,
   this->progress(proportionDone, progMessage.str());
 }
 
+/**
+ * Given a spectrum index, find the corresponding detector ID
+ * @param ws :: [input] Workspace to look in
+ * @param spectrumIndex :: [input] Spectrum index to look up
+ * @returns :: detector ID corresponding to spectrum index
+ * @throws std::runtime_error if it can't find the detector ID
+ */
+size_t CalMuonDetectorPhases::getDetectorIDFromSpectrumIndex(
+    const API::MatrixWorkspace_sptr &ws, const int spectrumIndex) {
+  const auto detidMap = ws->getDetectorIDToWorkspaceIndexMap();
+  for (const auto &entry : detidMap) {
+    if (static_cast<int>(entry.second) == spectrumIndex) {
+      return entry.first;
+    }
+  }
+  // Couldn't find it
+  std::ostringstream error;
+  error << "Could not find detector ID corresponding to spectrum "
+        << spectrumIndex;
+  throw std::runtime_error(error.str());
+}
+
```

Comment:
Good idea, that would be a better way. The function here is badly named too, it should be called `getDetectorIDFromWorkspaceIndex` not from "spectrum index" - I'll change this.